File: 263. Ugly Number.cpp
class Solution {
public:
    // brute force: 19.95%, ok
    bool isUgly(int num) {
        if(num < 1) return false;
        while(num % 10 == 0) num /= 10;  // not necessary, just to speed up
        while(num % 2 == 0) num /= 2;
        while(num % 3 == 0) num /= 3;
        while(num % 5 == 0) num /= 5;
        return num == 1;
    }

    // dp: O(N)time & space, Memory Limit Exceeded for large num
    bool isUgly(int num) {
        if(num < 1) return false;
        vector<int> dp(num + 1, 0);
        dp[0] = dp[1] = 1;  // corner case
        for(int i = 2; i <= num; i++){
            if((i % 2) && (i % 3) && (i % 5)) continue;
            else{
                if(i % 5 == 0) dp[i] = dp[i / 5];
                else if(i % 3 == 0) dp[i] = dp[i / 3];
                else if(i % 2 == 0) dp[i] = dp[i / 2];
            }
        }
        return dp[num];
    }
};

File: 263. Ugly Number.py
class Solution(object):
    
    # brute force decomposition: O(log5 + log3 + log2) time, O(1) space
    def isUgly(self, num):
        """
        :type num: int
        :rtype: bool
        """
        if num <= 0:  # boundry check
            return False
        if num & (num - 1) == 0:  # power of 2
            return True
        factors = [2, 3, 5]
        while num != 1:
            decompose = False
            for f in factors:
                if num % f == 0:
                    decompose = True
                    num /= f
            if not decompose:
                return False
        return True
    
    # faster brute force decomposition: O(log5 + log3 + log2) time, O(1) space
    def isUgly(self, num):
        if num <= 0:  # boundry check
            return False
        for f in [5, 3]:
            while num != 1 and num % f == 0:
                num /= f
        if num == 1:
            return True
        return num & (num - 1) == 0 # power of 2
    
    # pythonic brute force decomposition: O(log5 + log3 + log2) time, O(1) space
    def isUgly(self, num):
        if num <= 0:  # boundry check
            return False
        for f in [5, 3, 2]:
            while num % f == 0:
                num /= f
        return num == 1
     